File: examples/java/JPA_Hubernate/services.java
public class PostPredicates {

    private static final QPost post = QPost.post;
    private static final QPostSection section = QPostSection.postSection;

    public static BooleanExpression filterByBlog(Long blogId) {
        return blogId == null ? null : post.blog.id.eq(blogId);
    }

    public static BooleanExpression filterByAuthors(List<Long> authorIds) {
        return (authorIds == null || authorIds.isEmpty()) ? null : post.author.id.in(authorIds);
    }

    public static BooleanExpression filterBySectionCode(String sectionCode, int offset) {
        if (sectionCode == null || sectionCode.isEmpty()) return null;

        var subQuery = JPAExpressions
            .select(section.post.id)
            .from(section)
            .where(section.sectionCode.eq(sectionCode));

        return offset > 0 ? post.id.notIn(subQuery) : post.id.in(subQuery);
    }

    public static BooleanExpression filterBySectionStatuses(List<SectionStatus> statuses) {
        if (statuses == null || statuses.isEmpty()) return null;

        return JPAExpressions
            .selectOne()
            .from(section)
            .where(
                section.post.id.eq(post.id)
                .and(section.permitComment.eq(true))
                .and(section.sectionStatus.in(statuses))
            )
            .exists();
    }
}

@Service
@RequiredArgsConstructor
public class PostService {

    private final PostRepository postRepository;

    public List<Post> getValidPosts(Long blogId, List<Long> authorIds, String sectionCode, int offset, String editTiming) {
        // 도메인 정책에 따른 상태 목록 생성
        List<SectionStatus> allowedStatuses = getAllowedStatuses(editTiming);

        // Predicate 조립
        BooleanExpression predicate = Expressions.allOf(
            PostPredicates.filterByBlog(blogId),
            PostPredicates.filterByAuthors(authorIds),
            PostPredicates.filterBySectionCode(sectionCode, offset),
            PostPredicates.filterBySectionStatuses(allowedStatuses)
        );

        int page = offset / 10;
        var pageable = PageRequest.of(page, 10, Sort.by(Sort.Direction.DESC, "postedAt"));

        return postRepository.findAll(predicate, pageable).getContent();
    }

    private List<SectionStatus> getAllowedStatuses(String editTiming) {
        return switch (editTiming) {
            case "검토중" -> List.of(SectionStatus.검토중, SectionStatus.검토완료, SectionStatus.게시됨);
            case "검토완료" -> List.of(SectionStatus.검토완료, SectionStatus.게시됨);
            default -> Collections.emptyList();
        };
    }
}

File: examples/python/SQLAlchemy/services.py
from models import Post
from repositories import PostRepository

class PostService:
    def __init__(self, repo: PostRepository):
        self.repo = repo

    def get_valid_posts(self, blog_id: int, data: dict):
        author_ids = (
            [int(a) for a in data.get("author_ids", "").split(",") if a]
            if data.get("author_ids")
            else []
        )
        section_code = data.get("section_code")
        offset = int(data.get("offset", 0))
        edit_timing = 0  # 검토중

        allowed_statuses = [0]
        match edit_timing:
            case "검토중":
                allowed_statuses.extend(
                    [SectionStatus.검토중, SectionStatus.검토완료, SectionStatus.게시됨]
                )
            case "검토완료":
                allowed_statuses.extend([SectionStatus.검토완료, SectionStatus.게시됨])
            case _:
                pass

        q = self.repo.get_base_query()
        q = self.repo.filter_by_blog(q, blog_id)
        q = self.repo.filter_by_authors(q, author_ids)
        q = self.repo.filter_by_section_code(q, section_code, offset)
        q = self.repo.filter_by_section_status(q, allowed_statuses)

        return q.order_by(Post.posted_at.desc()).offset(offset).limit(10).all()
